cpp/src/arrow/type.cc

```diff
@@ -478,6 +479,47 @@ std::string DictionaryType::ToString() const {
   return ss.str();
 }
 
+Result<bool> DictionaryType::ComparableWithoutUnification(
+    std::vector<const Array*> arrays) {
+  if (arrays.size() == 0) {
+    return Status::Invalid("need at least one input array");
+  }
+
+  for (auto a : arrays) {
+    if (a->type_id() != Type::DICTIONARY) {
+      return Status::TypeError("input arrays must be dictionary arrays");
+    }
+  }
+
+  for (auto a : arrays) {
+    // XXX should this be a false return instead of an error? Unification can
+    // reconcile differing index types, for example
+    if (!a->type()->Equals(arrays[0]->type())) {
+      return Status::TypeError("array types not all consistent ", *arrays[0]->type(),
+                               " vs ", *a->type());
+    }
+  }
```

Comment:
I think returning false is better. This is doing too much in one function IMHO

Reply:
Not sure what you mean; what is doing too much? Checking for type agreement seems necessary

Reply:
I mean that returning an error is going too far

Reply:
I was following `DictionaryType::Unify` which returns Error status under this condition
https://github.com/apache/arrow/blob/2ba0566/cpp/src/arrow/array/builder_dict.cc#L130-L133
I can change this to a DCHECK